singer_sdk/sinks/sql.py

```diff
@@ -116,6 +122,42 @@ def full_schema_name(self) -> str:
             schema_name=self.schema_name, db_name=self.database_name
         )
 
+    def conform_name(self, name: str, object_type: Optional[str] = None):
+        """Conform a stream property name to one suitable for the target system.
+
+        Transforms names to snake case by default, applicable to most common DBMSs'.
+        Developers may override this method to apply custom transformations
+        to database/schema/table/column names.
+
+        Returns:
+            The name transformed to snake case.
+        """
+        # strip non-alphanumeric characters, keeping -, _ and spaces
+        name = re.sub(r"[^a-zA-Z0-9_\-\.\s]", "", name)
+        # convert to snakecase
+        name = snakecase(name)
+        # remove leading digits
+        return name.lstrip(digits)
```

Comment:
@kgpayne Perhaps adding an alpha prefix instead of a replacing leading digits is preferable?

Suppose there's properties `1_users`, `2_users`, then they'd both be conformed to `_users`.

Wdyt?

Reply:
@edgarrmondragon I thought about this a bit over the weekend, and think a good general approach would be to implement a post-conform check and custom error on "conformed name clash". This will be safe for us as we add to the corpus of cases handled in the default implementation (happy to add an alpha prefix to the default transformations too) but also safe for users as they override the conform recipes for their own targets 🙂

Reply:
@kgpayne 

> implement a post-conform check and custom error on "conformed name clash".

That sounds like a good approach 👍. Do you think name conforming errors are always bugs, or is there potential for "user errors" that would cause them? I ask because if it's the former, then we should think of a way to direct users to report the bug directly in the connector repo.

(That maybe applies too for schema errors in REST taps, but it just occurred to me)

> happy to add an alpha prefix to the default transformations too

Yeah, if you think it's worth it go for it 😄

Reply:
> Do you think name conforming errors are always bugs, or is there potential for "user errors" that would cause them?

Just to chime in here, I think name collisions are probably an area we'll need to come back to in future iterations. I don't remember the exact column names, but GitHub, for example has a '+1' and '-1' column for the emotes. And an early implementation of tap-github were basically rendering them with the same name, since both '-' and '+' were rendered as special characters.

I'd be okay with the 'hard failure' behavior when collisions are detected or 'warn and suppress the latter field' behavior. We probably don't want to _only_ warn, since if we potentially send both properties to the same target node, that could lead to data corruption.

If a hard failure occurs, especially if we give a clear error message, the user will go through:

> Oh, '+1' and '-1' both evaluate to 'c_1'? I don't really need those columns so I'll deselect them upstream in the pipeline.

Or

> Oh, '+1' and '-1' both evaluate to 'c_1'? Let me alias them explicitly in the tap...

So - I slightly prefer a hard failure because it gives the user more control over resolution. (Warnings are often missed in the massive volumes of logs.)

Reply:
Agree that conforming names will inevitably cause clashes. As of now, the default implementation tries to mitigate some obvious clashes, and raises a `ConformedNameClashException` if any clashes are detected. Users are free to override `conform_name()` for their own targets, so I think hard-fail is the safest approach 🙂